### PR TITLE
Add search endpoint

### PR DESCRIPTION
**Pull request by michielderoos**
Search endpoint!  
This uses postgres' full text search capabilities and joins with credit_transfer and transfer_account to effectively replace the usage of the `transfer_account_api` and `credit_transfer_api`.  
When you hit the search API without a search string (I.e. `/api/v1/search/?search_string=&search_type=transfer_accounts`), it will behave just like the old endpoint (pagination parameters and all). When you add anything to `search_string` though, it'll do what it says on the tin and actually return search results. It was designed like this in order to serve as a drop-in replacement for the old `transfer_account` and `credit_transfer` APIs. Also note that the return values are pretty well identical to the previous ones as well.  
Below are some examples of how to use the endpoint (Note: All of these are done with an auth header as well):   
`GET /api/v1/search/?search_string=fra der&search_type=transfer_accounts`  
(Search `transfer_accounts` for `fra der` to match `Francine deRoos`)  
  
`GET /api/v1/search/?search_string=mich&search_type=credit_transfers`  
(Search `credit_transfers` for `mich` to match any transfer to or from `Michiel deRoos`)
  
`GET /api/v1/search/?search_string=%2B19021231234&search_type=credit_transfers`  
(Search `credit_transfers` for phone number`+19021231234` to match any transfer to or from `Michiel deRoos`)

### CODE REVIEW

**Comment by EDsCODE** [comment]
[Server Side Search](https://trello.com/c/hVzL3f3r/208-server-side-search)

File: config.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
 from web3 import Web3
 
-VERSION = '1.0.29'  # Remember to bump this in every PR
+VERSION = '1.0.30'  # Remember to bump this in every PR
```
Comment (tristanhcole) [inline config.py]: not a feature bump?

File: app/server/api/search_api.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            user_search_result = db.session.query(
+                db.distinct(SearchView.id),
+                SearchView,
+                # This ugly (but functional) multi-tscolumn ranking is a modified from Ben Smithgall's blog post
+                # https://www.codeforamerica.org/blog/2015/07/02/multi-table-full-text-search-with-postgres-flask-and-sqlalchemy/
+                db.func.max(db.func.full_text.ts_rank(
+                    db.func.setweight(db.func.coalesce(SearchView.tsv_email, ''), 'D')\
+                        .concat(db.func.setweight(db.func.coalesce(SearchView.tsv_phone, ''), 'A'))\
+                        .concat(db.func.setweight(db.func.coalesce(SearchView.tsv_first_name, ''), 'B'))\
+                        .concat(db.func.setweight(db.func.coalesce(SearchView.tsv_last_name, ''), 'B')),
```
Comment (tristanhcole) [inline app/server/api/search_api.py]: I feel like the ability to also search for the following items would be very useful...

- Public Serial Number
- Location
- Blockchain Address
- And if possible, _any other custom attribute_ (i.e. Bio)

Not sure how hard that would be to add though!

File: test/functional/app/api/test_search_api.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
+"""
+This file tests search_api.py.
+"""
+import json, pytest
+from time import sleep
+from server.utils.auth import get_complete_auth_token
+from server.utils.user import create_transfer_account_user
+from server import db
+
+def test_transfer_usage_api_get(test_client, complete_admin_auth_token, create_organisation):
```
Comment (tristanhcole) [inline test/functional/app/api/test_search_api.py]: this test looks super clean and basically perfect, BUT name of test needs to be renamed `test_transfer_usage_api_get` 👀 and doc string  fixed

File: app/migrations/versions/1c4c356aa44c_.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+                to_tsvector(u.email) AS tsv_email,
+                to_tsvector(u._phone) AS tsv_phone,
+                to_tsvector(u.first_name) AS tsv_first_name,
+                to_tsvector(u.last_name) AS tsv_last_name
+            FROM "user" u
+        );
+    '''))
+
+    op.create_index(op.f('ix_search_view_id'), 'search_view', ['id'], unique=True)
+    op.create_index(op.f('ix_tsv_email'), 'search_view', ['tsv_email'], postgresql_using='gin')
```
Comment (tristanhcole) [inline app/migrations/versions/1c4c356aa44c_.py]: With a DB up to date with latest master migration, I got the following error when running `python manage.py db upgrade`

```
INFO  [alembic.runtime.migration] Running upgrade fc2eeba8e8e8 -> 1c4c356aa44c, empty message
Traceback (most recent call last):
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 552, in do_execute
    cursor.execute(statement, parameters)
psycopg2.ProgrammingError: column "ix_tsv_email" does not exist


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "manage.py", line 44, in <module>
    manager.run()
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/flask_script/__init__.py", line 417, in run
    result = self.handle(argv[0], argv[1:])
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/flask_script/__init__.py", line 386, in handle
    res = handle(*args, **config)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/flask_script/commands.py", line 216, in __call__
    return self.run(*args, **kwargs)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/flask_migrate/__init__.py", line 259, in upgrade
    command.upgrade(config, revision, sql=sql, tag=tag)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/alembic/command.py", line 279, in upgrade
    script.run_env()
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/alembic/script/base.py", line 475, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/alembic/util/pyfiles.py", line 98, in load_python_file
    module = load_module_py(module_id, path)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/alembic/util/compat.py", line 174, in load_module_py
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 205, in _call_with_frames_removed
  File "migrations/env.py", line 91, in <module>
    run_migrations_online()
  File "migrations/env.py", line 83, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/alembic/runtime/environment.py", line 846, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/alembic/runtime/migration.py", line 365, in run_migrations
    step.migration_fn(**kw)
  File "/Users/tristancole/PycharmProjects/SempoBlockchain/app/migrations/versions/1c4c356aa44c_.py", line 53, in upgrade
    op.create_index(op.f('ix_tsv_email'), 'search_view', ['ix_tsv_email'], postgresql_using='gin')
  File "<string>", line 8, in create_index
  File "<string>", line 3, in create_index
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/alembic/operations/ops.py", line 975, in create_index
    return operations.invoke(op)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/alembic/operations/base.py", line 345, in invoke
    return fn(self, operation)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/alembic/operations/toimpl.py", line 88, in create_index
    operations.impl.create_index(idx)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/alembic/ddl/impl.py", line 276, in create_index
    self._exec(schema.CreateIndex(index))
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/alembic/ddl/impl.py", line 134, in _exec
    return conn.execute(construct, *multiparams, **params)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 988, in execute
    return meth(self, multiparams, params)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/sqlalchemy/sql/ddl.py", line 72, in _execute_on_connection
    return connection._execute_ddl(self, multiparams, params)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1050, in _execute_ddl
    compiled,
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1253, in _execute_context
    e, statement, parameters, cursor, context
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1473, in _handle_dbapi_exception
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 398, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 152, in reraise
    raise value.with_traceback(tb)
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1249, in _execute_context
    cursor, statement, parameters, context
  File "/Users/tristancole/VirtualEnvs/SempoBlockchain_env/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 552, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (psycopg2.ProgrammingError) column "ix_tsv_email" does not exist

[SQL: CREATE INDEX ix_tsv_email ON search_view USING gin (ix_tsv_email)]
(Background on this error at: http://sqlalche.me/e/f405)
```

File: test/functional/app/api/test_search_api.py (hunk 3 above)
Comment (michielderoos) [reply test/functional/app/api/test_search_api.py]: Don't know how I missed that!

File: app/server/api/search_api.py (hunk 2 above)
Comment (michielderoos) [reply app/server/api/search_api.py]: Added serial/address/location. I think indexing custom attributes would be a bit trickier. Should we spin that off into another ticket?

File: app/migrations/versions/1c4c356aa44c_.py (hunk 4 above)
Comment (michielderoos) [reply app/migrations/versions/1c4c356aa44c_.py]: Fully nuked my local database, and all the migrations ran fine, I'm also able to downgrade and upgrade on/off this migration freely. I _did_ just upgrade this migration in the last batch of changes though, so perhaps if you try again it'll work now.

File: app/server/api/search_api.py (hunk 2 above)
Comment (tristanhcole) [reply app/server/api/search_api.py]: Yeah, I think keep custom attribute separate if it's to hard

**Comment by michielderoos** [comment]
@enjeyw   
  
> * when I tried searching via phone number, it didn't work for some reason. Possibly because phone is actually on the user table?

The phone table _is_ being indexed, and I can search it. I think what may have been happening has to do with URL encoding. When searching for a phone number with a plus, you'll have to URL encode the parameters (I.e. for `+1902`, use `search_string=%2B1902` instead of `search_string=+1902`

File: app/migrations/versions/1c4c356aa44c_.py (hunk 4 above)
Comment (tristanhcole) [reply app/migrations/versions/1c4c356aa44c_.py]: Seems to work now

**Comment by tristanhcole** [comment]
> @enjeyw
> 
> > * when I tried searching via phone number, it didn't work for some reason. Possibly because phone is actually on the user table?
> 
> The phone table _is_ being indexed, and I can search it. I think what may have been happening has to do with URL encoding. When searching for a phone number with a plus, you'll have to URL encode the parameters (I.e. for `+1902`, use `search_string=%2B1902` instead of `search_string=+1902`

You might also need to add `&org=2` or whatever org ID you're querying

**Comment by tristanhcole** [comment]
being able to search for phone numbers without country code would be good to implement once this is PR is merged in https://trello.com/c/XvUgYs4c/268-optional-country-codes

**Pull request merged**